### PR TITLE
Feature screen three domain

### PR DESCRIPTION
**Pull request by anarahari**
- Created EpisodeRepository interface to get episode details

- Created GetEpisodeDetailsUseCase to fetch the episode details response from repository

- Added unit test cases for GetEpisodeDetailsUseCase to fetch episode details on success and failure scenarios.

- Added characters variable to Episode model to fetch list of characters.

- Changes for lintchecks

### CODE REVIEW

File: app/src/main/java/com/compose/rickymortypoc/MainActivity.kt [hunk 1]
```diff
@@ -21,12 +18,12 @@ class MainActivity : ComponentActivity() {
...
-                    ViewModelProvider(
-                        LocalContext.current as ComponentActivity,
-                        viewModelFactoryProvider
-                    )[CharacterViewModel::class.java]
-                CharacterListScreen(characterViewModel)
+//                val characterViewModel: CharacterViewModel =
+//                    ViewModelProvider(
+//                        LocalContext.current as ComponentActivity,
+//                        viewModelFactoryProvider
+//                    )[CharacterViewModel::class.java]
```
Comment (shreyansh-mm) [inline app/src/main/java/com/compose/rickymortypoc/MainActivity.kt]: Remove the commented code

File: data/src/main/java/com/compose/data/datamapper/DataMapper.kt [hunk 2]
```diff
@@ -1,17 +1,62 @@
...
+            species = it.character?.species,
+            status = it.character?.status,
+            origin = mapToOriginModel(it.character?.origin),
+            location = mapToLocationModel(it.character?.location),
+            episodes = mapToEpisodeModel(it.character?.episode),
+        )
+    }
+}
+
+fun mapToOriginModel(origin: GetCharacterDetailsQuery.Origin?): Origin {
```
Comment (shreyansh-mm) [inline data/src/main/java/com/compose/data/datamapper/DataMapper.kt]: private

File: data/src/main/java/com/compose/data/datamapper/DataMapper.kt [hunk 3]
```diff
@@ -1,17 +1,62 @@
...
+    return Origin(
+        name = origin?.name,
+        dimension = origin?.dimension
+    )
+}
+
+fun mapToLocationModel(location: GetCharacterDetailsQuery.Location?): Location {
```
Comment (shreyansh-mm) [inline data/src/main/java/com/compose/data/datamapper/DataMapper.kt]: private

File: data/src/main/java/com/compose/data/datamapper/DataMapper.kt [hunk 4]
```diff
@@ -1,17 +1,62 @@
...
+    return Location(
+        name = location?.name,
+        dimension = location?.dimension,
+        id = location?.id
+    )
+}
+
+fun mapToEpisodeModel(episodes: List<GetCharacterDetailsQuery.Episode?>?): List<Episode>? {
```
Comment (shreyansh-mm) [inline data/src/main/java/com/compose/data/datamapper/DataMapper.kt]: private

File: data/src/main/java/com/compose/data/repository/CharacterRepositoryImpl.kt [hunk 5]
```diff
@@ -21,4 +23,13 @@ internal class CharacterRepositoryImpl @Inject constructor(private val apolloCli
...
         }
     }
+
+    override suspend fun getCharacterDetails(id: String): Character? {
+        return try {
+            val response = apolloClient.query(GetCharacterDetailsQuery(id)).execute()
+            getCharacterDetailsQueryToCharacterModel(response.data)
+        } catch (ex: Exception) {
+            null
+        }
```
Comment (shreyansh-mm) [inline data/src/main/java/com/compose/data/repository/CharacterRepositoryImpl.kt]: Exception should not be consumed, it should be propagated forward.
So we can show appropriate error to use.

File: domain/src/test/java/com/compose/domain/usecase/GetCharacterDetailsUseCaseTest.kt [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+    private lateinit var SUT: GetCharacterDetailsUseCase
+    private val testDispatcher = StandardTestDispatcher()
+    @RelaxedMockK
+    private lateinit var repository: CharacterRepository
+
+    @Before
+    fun setUp() {
+        MockKAnnotations.init(this)
+        Dispatchers.setMain(testDispatcher)
```
Comment (shreyansh-mm) [inline domain/src/test/java/com/compose/domain/usecase/GetCharacterDetailsUseCaseTest.kt]: Test should work fine without this line

File: domain/src/test/java/com/compose/domain/usecase/GetCharacterDetailsUseCaseTest.kt [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+            val characterId = "-1"
+            val exception = RuntimeException("Error fetching character details")
+            coEvery { repository.getCharacterDetails(characterId) } throws exception
+            val result = SUT.invoke(characterId)
+            assertEquals("Error fetching character details", result.last().message)
+        }
+
+    @After
+    fun tearDown() {
+        Dispatchers.resetMain()
```
Comment (shreyansh-mm) [inline domain/src/test/java/com/compose/domain/usecase/GetCharacterDetailsUseCaseTest.kt]: this should also be not needed

File: presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+) {
+    LaunchedEffect(characterId) {
+        viewModel.getCharacterDetails(characterId)
+    }
+    val characterState by viewModel.characterDetailsState.collectAsStateWithLifecycle()
+    CharacterDetailsItems(characterState, toolbarTitle, onBackButtonPressed = onBackButtonPressed)
+}
+
+@Composable
+fun CharacterDetailsItems(
```
Comment (shreyansh-mm) [inline presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt]: private

File: presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+        }
+
+        characterState.data?.let {
+            DisplayCharacterDetailsData(modifier, it as Character)
+        }
+    }
+}
+
+@Composable
+fun DisplayCharacterDetailsData(
```
Comment (shreyansh-mm) [inline presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt]: private

File: presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+        CustomAppBar(toolbarTitle, onBackButtonPressed = onBackButtonPressed)
+        if (characterState.isLoading) {
+            CircularProgressIndicator(Modifier.align(Alignment.CenterHorizontally))
+        }
+
+        if (characterState.error.isNotEmpty()) {
+            Text(
+                text = characterState.error.toString(),
+                color = MaterialTheme.colorScheme.error,
+                modifier = Modifier
```
Comment (shreyansh-mm) [inline presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt]: You don't need to pass empty modifier

File: presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt [hunk 11]
```diff
@@ -0,0 +1,335 @@
...
+        }
+    }
+}
+
+@Composable
+fun DisplayCharacterDetailsData(
+    modifier: Modifier = Modifier,
+    character: Character,
+) {
+    Column(modifier.verticalScroll(rememberScrollState())) {
```
Comment (shreyansh-mm) [inline presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt]: let's move `rememberScrollState()` to a variable

File: presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt [hunk 12]
```diff
@@ -0,0 +1,335 @@
...
+        ) {
+            AsyncImage(
+                model = character.image,
+                contentDescription = character.name,
+                modifier = Modifier
+                    .fillMaxWidth()
+                    .aspectRatio(1f)
+            )
+        }
+        Spacer(modifier.padding(top = MaterialTheme.dimens.paddingExtraLarge))
```
Comment (shreyansh-mm) [inline presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt]: Modifier coming from args should only apply to parent composable.

File: presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt [hunk 13]
```diff
@@ -0,0 +1,335 @@
...
+        Text(
+            text = character.name.orEmpty(),
+            fontWeight = FontWeight.Bold,
+            style = MaterialTheme.typography.headlineLarge,
+            color = Color.Black,
```
Comment (shreyansh-mm) [inline presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt]: This should come from `MaterialTheme`

File: presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt [hunk 14]
```diff
@@ -0,0 +1,335 @@
...
+            modifier = modifier.padding(start = MaterialTheme.dimens.paddingExtraLarge)
```
Comment (shreyansh-mm) [inline presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt]: New modifier should be used

File: presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt [hunk 15]
```diff
@@ -0,0 +1,335 @@
...
+        )
+        CharacterStatus(modifier, character)
```
Comment (shreyansh-mm) [inline presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt]: This modifier should not be passed forward

File: presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt [hunk 16]
```diff
@@ -0,0 +1,335 @@
...
+    }
+}
+
+@Composable
+fun CharacterStatus(modifier: Modifier, character: Character) {
```
Comment (shreyansh-mm) [inline presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt]: private

File: presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt [hunk 17]
```diff
@@ -0,0 +1,335 @@
...
+        modifier = modifier.padding(
+            start = MaterialTheme.dimens.paddingExtraLarge,
+            top = MaterialTheme.dimens.paddingLarge
+        )
+    ) {
+        Text(
+            text = stringResource(R.string.status),
+            fontSize = 16.sp,
+            fontWeight = FontWeight.Bold,
+            color = Color.Black
```
Comment (shreyansh-mm) [inline presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt]: all these values can come from `textStyle` / `typography`

File: presentation/src/main/java/com/compose/presentation/screens/FirstScreen.kt [hunk 18]
```diff
@@ -15,29 +16,43 @@ import androidx.compose.material3.CircularProgressIndicator
...
+    modifier: Modifier = Modifier,
+    onNavigateCharacterDetails: (String) -> Unit
+) {
+    var areCharactersFetched by rememberSaveable { mutableStateOf(false) }
+    LaunchedEffect(Unit) {
+        if (!areCharactersFetched) {
+            areCharactersFetched = true
+            viewModel.getCharacters()
+        }
+    }
```
Comment (shreyansh-mm) [inline presentation/src/main/java/com/compose/presentation/screens/FirstScreen.kt]: our UI layer needs to be as dumb as possible, this logic should not sit here

**Review by shreyansh-mm — CHANGES_REQUESTED** [review]

File: presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt [hunk 19]
```diff
@@ -0,0 +1,310 @@
...
+        CharacterSpecies(modifier, character)
+        CharacterGender(modifier, character)
+        CharacterOrigin(modifier, character)
+        CharacterLocation(modifier, character)
+        DisplayEpisodes(modifier, character)
+    }
+}
+
+@Composable
+private fun CharacterStatus(modifier: Modifier = Modifier, character: Character) {
```
Comment (shubhamsinghmutualmobile) [inline presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt]: This composable should only ask for the `status` variable

File: presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt [hunk 20]
```diff
@@ -0,0 +1,310 @@
...
+        Text(
+            text = character.status.orEmpty().uppercase(),
+            style = MaterialTheme.typography.titleMedium,
+            modifier = modifier.padding(start = MaterialTheme.dimens.paddingSmall),
+        )
+    }
+}
+
+@Composable
+private fun CharacterSpecies(modifier: Modifier, character: Character) {
```
Comment (shubhamsinghmutualmobile) [inline presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt]: This composable should only ask for the `species` variable

File: presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt [hunk 21]
```diff
@@ -0,0 +1,310 @@
...
+        Text(
+            text = character.species.orEmpty().uppercase(),
+            style = MaterialTheme.typography.titleMedium,
+            modifier = Modifier.padding(start = MaterialTheme.dimens.paddingSmall)
+        )
+    }
+}
+
+@Composable
+private fun CharacterGender(modifier: Modifier, character: Character) {
```
Comment (shubhamsinghmutualmobile) [inline presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt]: This composable should only ask for the `gender` variable

File: presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt [hunk 22]
```diff
@@ -0,0 +1,310 @@
...
+        Text(
+            text = character.gender.orEmpty().uppercase(),
+            style = MaterialTheme.typography.titleMedium,
+            modifier = Modifier.padding(start = MaterialTheme.dimens.paddingSmall)
+        )
+    }
+}
+
+@Composable
+private fun CharacterOrigin(modifier: Modifier, character: Character) {
```
Comment (shubhamsinghmutualmobile) [inline presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt]: This composable should only ask for the `origin` variable

File: presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt [hunk 23]
```diff
@@ -0,0 +1,310 @@
...
+        Text(
+            text = character.origin.dimension.orEmpty().uppercase(),
+            style = MaterialTheme.typography.titleMedium,
+            modifier = modifier.padding(start = MaterialTheme.dimens.paddingSmall),
+        )
+    }
+}
+
+@Composable
+private fun CharacterLocation(modifier: Modifier, character: Character) {
```
Comment (shubhamsinghmutualmobile) [inline presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt]: This composable should only ask for the `location` variable

File: presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt [hunk 24]
```diff
@@ -0,0 +1,310 @@
...
+        Text(
+            text = character.location.dimension.orEmpty().uppercase(),
+            style = MaterialTheme.typography.titleMedium,
+            modifier = Modifier.padding(start = MaterialTheme.dimens.paddingSmall)
+        )
+    }
+}
+
+@Composable
+private fun DisplayEpisodes(modifier: Modifier, character: Character) {
```
Comment (shubhamsinghmutualmobile) [inline presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt]: This composable should only ask for the `episodes` variable

File: presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt [hunk 25]
```diff
@@ -0,0 +1,310 @@
...
+        modifier = Modifier.fillMaxWidth(),
+        contentPadding = PaddingValues(all = MaterialTheme.dimens.paddingLarge),
+        horizontalArrangement = Arrangement.spacedBy(MaterialTheme.dimens.paddingSmall)
+    ) {
+        character.episodes?.size?.let {
+            items(
+                it,
+                key = { item -> character.episodes?.get(item)?.id!! }
+            ) { index ->
+                EpisodesItem(modifier, episode = character.episodes?.get(index)!!)
```
Comment (shubhamsinghmutualmobile) [inline presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt]: Avoid using the double bang operator (`!!`) whenever you can

**Review by shubhamsinghmutualmobile — APPROVED** [review]

**Pull request merged**